librb/src/openssl.c

```diff
@@ -392,28 +392,28 @@ rb_setup_ssl_server(const char *const certfile, const char *keyfile,
 	}
 	else
 	{
-		FILE *const dhf = fopen(dhfile, "r");
-		DH *dhp = NULL;
+		BIO *const dhf = BIO_new_file(dhfile, "r");
+		EVP_PKEY* dhp = NULL;
 
 		if(dhf == NULL)
 		{
-			rb_lib_log("%s: fopen ('%s'): %s", __func__, dhfile, strerror(errno));
+			rb_lib_log("%s: BIO_new_file ('%s'): %s", __func__, dhfile,
+			           rb_ssl_strerror(rb_ssl_last_err()));
 		}
-		else if(PEM_read_DHparams(dhf, &dhp, NULL, NULL) == NULL)
+		else if(PEM_read_bio_Parameters(dhf, &dhp) == NULL)
 		{
-			rb_lib_log("%s: PEM_read_DHparams ('%s'): %s", __func__, dhfile,
+			rb_lib_log("%s: PEM_read_bio_Parameters ('%s'): %s", __func__, dhfile,
 			           rb_ssl_strerror(rb_ssl_last_err()));
-			fclose(dhf);
+			BIO_free(dhf);
 		}
 		else
 		{
 			SSL_CTX_set_tmp_dh(ssl_ctx_new, dhp);
```

Comment:
According to [the documentation](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_tmp_dh.html), this function is also deprecated.

It provides a suitable function to achieve the same result, `SSL_CTX_set0_tmp_dh_pkey(SSL_CTX *, EVP_PKEY *)`.

However, given that this is a `set0` function (the OpenSSL API convention for a function that assumes ownership of an object without incrementing its reference count), the object must *NOT* be freed (with `EVP_PKEY_free(3ssl)`) if this function succeeds. This function is not available in version 1.1, so it will need some `#if OPENSSL_VERSION_NUMBER` selection logic like the rest of the file.

Reply:
How annoying ! I think I have that cleaned up now. Still working on setting up a test server to test challenge on. Should I mark this as a draft until then?